common/src/main/java/io/pravega/common/tracing/RequestTracker.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.tracing;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This class is a helper for multiple classes to get the client-generated id for a request (requestId) based on the
+ * available information of the request itself (requestDescriptor) in multiple methods. The objective is to cache such
+ * [requestDescriptor, requestId] pairs to we allow classes within a component (Controller, Segment Store) to also log
+ * the client-generated ids, making it easier to trace the lifecycle of a request across multiple logs.
+ */
+@Slf4j
+public final class RequestTracker {
+
+    private static final RequestTracker INSTANCE = new RequestTracker();
+    private static final String INTER_FIELD_DELIMITER = "-";
+
+    private final Cache<String, List<Long>> ongoingRequests;
+
+    private RequestTracker() {
+        // Clean request tags after a certain amount of time.
+        ongoingRequests = CacheBuilder.newBuilder()
+                                      .maximumSize(10000)
+                                      .expireAfterWrite(10, TimeUnit.MINUTES)
+                                      .build();
+    }
+
+    public static RequestTracker getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Creates a request descriptor or key to locate the client request id.
+     *
+     * @param requestInfo Fields to form the request descriptor.
+     * @return Request descriptor.
+     */
+    public static String buildRequestDescriptor(String...requestInfo) {
+        return Stream.of(requestInfo).collect(Collectors.joining(INTER_FIELD_DELIMITER));
+    }
+
+    public static String buildRequestDescriptor(String first, String[] rest) {
+        return buildRequestDescriptor(Lists.asList(first, rest).toArray(new String[rest.length]));
```

Comment:
You're creating an entirely new array. You don't need to.

Reply:
This method has been removed with changes in this PR.